### PR TITLE
Fixes #15739: Compact REPL printouts of an `Associative` use brackets when appropriate

### LINKED ISSUE #15739: `Dict` REPL output lacks parentheses
```Julia
julia> a1 = Dict(('a'=>1) => ('b'=>2))
Dict('a'=>1=>'b'=>2)
```
The way this `Dict` is output lacks at least one set of parentheses. If we input this expression ...
```
julia> Dict('a'=>1=>'b'=>2)
Dict('a'=>1=>('b'=>2))
```
... we get a different result.

### PR DESCRIPTION
**Pull request by Tetralux**
Displaying a `Dict` in compact format manually prints out the pairs in the form `key=>value` which is problematic if either the key or value is a `Pair`, itself.

This PR defers the printing of the ~~keys and the values~~ key-value pairs to `show`.

E.g:

On master:

``` julia
julia> Dict( (1=>2)=>(3=>10), (2=>4)=>(6=>20) )
Dict(2=>4=>6=>20,1=>2=>3=>10)
```

With this PR: 

``` julia
julia> Dict( (1=>2)=>(3=>10), (2=>4)=>(6=>20) )
Dict((2=>4)=>(6=>20),(1=>2)=>(3=>10))
```

### CODE REVIEW

**Comment by tkelman** [comment]
Thanks! Looks like the display order differs on 32 bit (different hashing results) which is causing the tests to fail on those platforms.

**Comment by H-225** [comment]
@tkelman Oops! And here's take two. Or at least, when the Travis queue gets to it.

File: test/dict.jl [hunk 1]
```diff
@@ -294,6 +294,18 @@ for d in (Dict("\n" => "\n", "1" => "\n", "\n" => "2"),
...
+# Issue #15739 - Compact REPL printouts of an `Associative` use brackets when appropriate
+let d = Dict((1=>2) => (3=>45), (3=>10) => (10=>11))
+    buf = IOBuffer()
+    showcompact(buf, d)
+
+    # Ordering differs on CPU bitness, so build up the output for comparision
+    # each time.
+    expectedstr = join([string(x) for x in d], ",")
+
+    @test bytestring(buf) == "Dict($expectedstr)"
```
Comment (Keno) [inline test/dict.jl]: Why not check whether it contains both expected strings. If somebody were to say break printing of Pairs, this would still pass.

Comment (H-225) [inline test/dict.jl]: @Keno A fair point. I'll have it check for all the strings we care about.

**Comment by Keno** [comment]
LGTM, please squash the commits if you don't mind.

**Comment by Keno** [comment]
Actually, let's try github's newfangled squash/merge. Hold on.

**Comment by tkelman** [comment]
It annoyingly drops the commit messages for later commits, and I kind of miss the "merge PR" reference in the log but it's nice that github at least gives the option now.

**Comment by H-225** [comment]
@tkelman @Keno Thank you! :smile:

**Comment by H-225** [comment]
@Keno Is it wise to rebase this branch and/or reuse it to fix the problem at https://github.com/JuliaLang/julia/issues/15739 - or shall I make a new PR?

**Comment by Keno** [comment]
Please make a new PR.

**Comment by vtjnash** [comment]
to expand on what Keno said, you might be able to use the same branch name for a new PR, but github doesn't allow reopening a PR once it has been merged

**Comment by H-225** [comment]
@vtjnash Good to know. Thanks.

**Pull request merged**